planning/behavior_path_planner/src/planner_manager.cpp

```diff
@@ -341,73 +342,101 @@ std::pair<SceneModulePtr, BehaviorModuleOutput> PlannerManager::runCandidateModu
     return SceneModulePtr();
   }();
 
+  /**
+   * register candidate modules.
+   */
   updateCandidateModules(executable_modules, module_ptr);
 
   return std::make_pair(module_ptr, results.at(module_ptr->name()));
 }
 
 BehaviorModuleOutput PlannerManager::runApprovedModules(const std::shared_ptr<PlannerData> & data)
 {
-  BehaviorModuleOutput output = getReferencePath(data);  // generate root reference path.
+  std::unordered_map<std::string, BehaviorModuleOutput> results;
 
-  bool remove_after_module = false;
+  BehaviorModuleOutput output = getReferencePath(data);
+  results.emplace("root", output);
 
-  for (auto itr = approved_module_ptrs_.begin(); itr != approved_module_ptrs_.end();) {
-    const auto & name = (*itr)->name();
+  /**
+   * execute all approved modules.
+   */
+  for (const auto & module_ptr : approved_module_ptrs_) {
+    output = run(module_ptr, data, output);
+    results.emplace(module_ptr->name(), output);
+  }
 
-    stop_watch_.tic(name);
+  /**
+   * pop waiting approve module. push it back candidate modules.
+   */
+  {
+    const auto waiting_approval_modules = [this](const auto & m) { return m->isWaitingApproval(); };
 
-    // if one of the approved modules changes to waiting approval, remove all behind modules.
-    if (remove_after_module) {
-      deleteExpiredModules(*itr);
-      itr = approved_module_ptrs_.erase(itr);
-      processing_time_.at(name) += stop_watch_.toc(name, true);
-      continue;
+    const auto itr = std::find_if(
+      approved_module_ptrs_.begin(), approved_module_ptrs_.end(), waiting_approval_modules);
+
+    if (itr != approved_module_ptrs_.end()) {
+      clearCandidateModules();
+      candidate_module_ptrs_.push_back(*itr);
```

Comment:
Just for confirmation:
What if there is more than one `waiting_approval_modules`?
For example lane change and avoidance both need approval.

From this code, my understanding is, if avoidance is first found, then `avoidance` will be pushed back to `candidate_module_ptrs`, and once `erase` is called, then both `avoidance` and `lane change` will be removed.

I still haven't checked the actual behavior, but from my understanding, for `enable_simultaneous_execution_as_candidate_module`, this might cause the code to run twice, in order to find lane change. Is my understanding correct?

Reply:
Thanks.

> From this code, my understanding is, if avoidance is first found, then avoidance will be pushed back to candidate_module_ptrs, and once erase is called, then both avoidance and lane change will be removed.

You're right, but this code is a little bit unintentionally. It should not call clearCandidateModules() twice. (I have to fix it.)

I want to reset candidate modules `candidate_modules_ptrs_` once before some approved modules pop and push them back to `candidate_modules_ptrs_`. 

For example, `candidate_modules_ptrs_` has ext LC right, and `approved_module_ptrs_` has avoidance, normal LC left.

C: `candidate_modules_ptrs_`
A: `approved_module_ptrs_`

Therefore, the expected behavior is,

1. remove and reset all modules in `candidate_modules_ptrs_`. (C: **EMPTY**, A: avoidance, normal LC left)
2. check waiting approval modules in `approved_module_ptrs_`. (both avoidance and normal LC left is waiting approval)
3. pop waiting approval modules in `approved_module_ptrs_`.
4. push them back to `candidate_modules_ptrs_`. (C: avoidance, normal LC left, A: **EMPTY**)

The reason why it should reset `candidate_modules_ptrs_` is that I think it is better to go back to square one and start candidate module's process again in order to remove residual modules that don't consider latest `approved_module_ptrs_` condition. But, there might be ways to work them well without reset.

Reply:
> You're right. If avoidance module is waiting approval and is found first in std::find_if, the module and next module (lane change) are removed in erase, even if the next module is also waiting aproval since the lane change module's input (= avoidance module output) maybe change significantly.

Now I understand. Thank you for the explanation. 
Since the intention is a little bit not obvious, I think it is better to put the explanation here as comment.

Reply:
Thanks. I'll do that :+1:

Reply:
Added comment in https://github.com/autowarefoundation/autoware.universe/pull/3382/commits/66096a0af2bca46194cb7d699d9c90663edff8fe.